### ISSUE TITLE
[suggestion] DAEMON as saga's default mode

### ISSUE THREAD

**Issue by FourwingsY** [react-boilerplate/react-boilerplate#2121]
from #2021 

I thought the problem is on saga's default mode: RESTART_ON_REMOUNT

regarding to https://github.com/facebook/react/issues/11106
garaeon saids:

> Regarding componentWillUnmont.
> 
> The official docs explicitly recommend use of this hook as cleanup, so this will probably break a lot of people unless changed/docs update/some workaround is recommended.
> 
> The docs are talking about clean up on the component level. For example listeners added by a component.
> 
> What you're talking about is cleaning up global state. This is inherently messy and can get very complicated. You're welcome to file an issue or PR in the docs repo to clarify it.
> 
> fire a cleanup action/set a flag to reset a part of the Redux state tree before showing another component that also uses that part of the state tree but needs it to be re-initialized before render
> 
> I've written such code myself before but always regretted it. This is brittle. I recommend refactoring away from patterns where you need this.
> 
> The easiest way to do so is to not reuse the same part of the tree in such a way. Think about it: components are generally expected to be independent. What happens if you render two such "exclusive" components at the same time? They will conflict. This doesn't seem like the right component thinking.
> 
> I understand it is more code to avoid such patterns, but in my experence replacing a single state tree branch with an object or an array was enough. For example you can key it by ID (even a dynamically generated one), and then instead of tearing down the old one in time for the new component mounting, you can just point the new component to the branch with new ID, while keeping the old one alive. Then eventually you can emit an action that cleans up unused state branches and only leaves the one with the most recent ID.

So ejecting sagas with componentWillUnmount is unnecessary. Sagas are on global-level, and It should not controlled by component-level actions. Sagas should only injected just like reducers. 

I think we should remove RESTART_ON_REMOUNT options, and set DAEMON as default.
Actually mods are not necessary, if we do this.

How do you guys think about?

**Comment by blling** [comment]
Saga injection issue is so sneaky, i think we need experts :)
cc @Dattaya @justingreenberg

**Comment by svitekpavel** [comment]
I just run into similar issue as in #2132. I was trying to use saga from other container to update data through API. The problem was that the saga was ejected in the middle of the API call and I didn't see any information in the console.

To the topic:
I don't think it's necessary to make all sagas DAEMONS, but I would suggest two things:

1. Add a global daemon sagas into `containers/App/saga.js` and mount it in `containers/App/index.js` like this:
```
import { DAEMON } from 'utils/constants';
const withSaga = injectSaga({ key: 'appDaemons', saga, mode: DAEMON });
export default compose(
  withSaga,
)(App);
```
2. Add more info into documentation

That would IMHO give developers a better understanding and realization that there's something like "daemon" sagas :-)

**Comment by FourwingsY** [comment]
I have a question. for who saw this comment.
"When do you need saga ejection and remount?"

**Comment by heyrict** [comment]
I use websocket along with redux saga for real-time updates.
The webpage should listen to notification updates all the time (run as DAEMON)
but post updates only when specific pages are mounted (run as RESTART_ON_REMOUNT)

though judging the current uri in saga is possible, RESTART_ON_REMOUNT is a much cleaner way.
@FourwingsY

**Comment by FourwingsY** [comment]
@heyrict hmm, does notification actions are intended to trigger post updates?

**Comment by heyrict** [comment]
Well, notification is a 'global' stuff that listens to updates when user is browsing the application (e.g. direct messages),while post update is a 'local' stuff that listens to updates when user is browsing a specific webpage(e.g. twitter new posts or "like" updates), which is not needed on other pages (e.g. settings or preferences).-------- 元のメッセージ --------件名: Re: [react-boilerplate/react-boilerplate] [suggestion] DAEMON as saga's default mode (#2121)From: 양현석 To: react-boilerplate/react-boilerplate Cc: heyrict ,Mention @heyrict hmm, does notification actions are intended to trigger post updates?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/react-boilerplate/react-boilerplate","title":"react-boilerplate/react-boilerplate","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/react-boilerplate/react-boilerplate"}},"updates":{"snippets":[{"icon":"PERSON","message":"@FourwingsY in #2121: @heyrict hmm, does notification actions are intended to trigger post updates?"}],"action":{"name":"View Issue","url":"https://github.com/react-boilerplate/react-boilerplate/issues/2121#issuecomment-375661104"}}}

**Comment by FourwingsY** [comment]
@heyrict Yes, that data or updates will not need for other pages. But what I want to talk about are redux and the state that redux is handling. **The state of redux** is on global level, not the component or page level. So what I think is both reducers and sagas should be treated as global level codes, even if it is only needed on single page.

**Comment by heyrict** [comment]
Yes, redux store is on a global level. But you may not want an unmounted component to take all actions to update the redux store **as a default** right?-------- 元のメッセージ --------件名: Re: [react-boilerplate/react-boilerplate] [suggestion] DAEMON as saga's default mode (#2121)From: 양현석 To: react-boilerplate/react-boilerplate Cc: heyrict ,Mention @heyrict Yes, that data or updates will not need for other pages. But what I want to talk about are redux and the state that redux is handling. The state of redux is on global level, not the component or page level. So what I think is both reducers and sagas should be treated as global level codes, even if it is only needed on single page.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/react-boilerplate/react-boilerplate","title":"react-boilerplate/react-boilerplate","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/react-boilerplate/react-boilerplate"}},"updates":{"snippets":[{"icon":"PERSON","message":"@FourwingsY in #2121: @heyrict Yes, that data or updates will not need for other pages. But what I want to talk about are redux and the state that redux is handling. **The state of redux** is on global level, not the component or page level. So what I think is both reducers and sagas should be treated as global level codes, even if it is only needed on single page."}],"action":{"name":"View Issue","url":"https://github.com/react-boilerplate/react-boilerplate/issues/2121#issuecomment-375720138"}}}

**Comment by FourwingsY** [comment]
@heyrict What I believe is: It depends on how you dispatch actions on your code, saga does not have to and need to be ejected with unmount. Preventing update store at such situation is developers responsibility. If some other components dispatches the action, and it updates unmounted-page-only-related state, It would be a wrong code design. 
If you have to eject saga for preventing updates of unmounted components, I think that is covering wrong component-action design with wrong saga-control design.

**Comment by howardya** [comment]
Just my 2 cents here:

It is the developer responsibility to make sure the actions are well specified and well targeted. In fact, it is the most important advice and the philosophy behind this boilerplate. So the argument that "we should not use DAEMON mode so that all the pages' saga start picking up unwanted actions" is same as "using a not intuitive approach to solve a not intuitive problem is the intuitive way".

So, I personally feel DAEMON mode is THE intuitive default mode.

**Comment by danielrs** [comment]
I agree with @FourwingsY and @howardya. I got started today with sagas, and using them in react-boilerplate and I intuitively thought that their behavior was similar to what `DAEMON` mode offers.

A common use case if a page with sub-components, each sub-component injecting the saga, as they are independent from one another. Everything is fine if you only use one component at a time, but if you instantiate more than one, you get multiple sagas running under the same key, which is far from intuitive.

Just my opinion.

**Comment by julienben** [comment]
Closing as this change has been implemented thanks to @howardya and #2430.
